website/content/functional-database.md

```diff
@@ -0,0 +1,298 @@
+non-existing arguments. Another way to phrase it, this is a constant expression.
+
+And as such, it might as well be defined as one:
+
+```js
+let my_var = 42;
+```
+
+In conventional programming languages[1], there is a difference between using
+`my_function()` and `my_var`. The first one evaluates the expression at the call
+site, while the second one evaluates it at the declaration site. It is also
+possible to express just `my_function`, a reference to a function that can be
+called without arguments.
+
+But if all your functions are pure, we generalize by saying that all three cases
+are equivalent.
+
+To make this work, let's say that `my_function` is "implicitly invoked" just as
+it would have been expressed as `my_function()`.
+
+Also, if we declare all functions to be pure it doesn't matter exactly _when_
+they are evaluated. So let's give the compiler the authority to make an informed
+guess about _when_ and make `my_var` and `my_function()` semantically
+equivalent.
+
+[1]: Let's say that conventional languages are first 15 from this list:
+https://survey.stackoverflow.co/2022/#most-popular-technologies-language-prof
+
+## Syntax, surly shinier
+
+First of all, lets change the function call to this:
+
+```
+(my_function arg1 arg2 arg3)
+```
+
+It may look strange because function name is within the parenthesis and there is
+no commas between arguments. But trust me, there are benefits to this syntax.
+
+Firstly, when the call has no arguments, the parenthesis can be omitted:
+
+```
+(my_function) == my_function
+```
+
+which feels very natural because of the similar behavior with expressions. As
+intended, function call with no argument and a plain reference to the function
+are expressed with the same syntax.
+
+Let's extend this behavior and allow allow bare function calls in a few places
+where they cannot become ambiguous:
+
+```
+# a list:
+[my_function arg1 arg2 arg3, my_function arg4 arg5 arg6]
+# a declaration:
+let res = my_function arg1 arg2 arg3
+```
+
+Secondly, I'd argue that currying looks natural.
+
+```
+let curry = my_function arg1 arg2
+let res = curry arg3
+
+# or inline:
+let res = (my_function arg1 arg2) arg3
+```
+
+Now let's go one step further and introduce a "pipe" operator. It applies left
+operand as an argument to the right operand:
+
+```
+arg3 | my_function arg1 arg2
+```
+
+This is especially useful for chaining function calls. If we declare `div`,
+`floor` and `mul` as common arithmetic functions, the function call syntax
+really starts to make sense:
+
+```
+12 | div 5 | floor | mul 5
+```
+
+## Running real relations
+
+Up to this point, we were talking language design without a clear justification.
+Language being cool is fine, but it is much more important that the language is
+actually the right tool for the job.
+
+The job, in the case of PRQL, is querying databases.
+
+Core unit of data here is a relation: an ordered set of rows, each of which
```

Comment:
> Core unit of data here is a relation: an ordered set of rows,

Is a relation an *ordered set of rows* or a *set of ordered tuples*?

I don't really know the answer but to me the statement seemed to imply that there was an inherent order in the data whereas my impression has always been that there are no order guarantees in SQL until you introduce an `ORDER BY` clause (cf #1363).

Reply:
Well there is an inherent order, it's just that for tables it's unknown and that all SQL SELECT statements without ORDER BY may change it.

So yes, in practice this is almost the same as relations being unordered, but with that definition, ORDER BY does not make much sense. As expressed in #1363, I'd prefer that PRQL does this better.

Also, I just realized that rows and tuples seem to be another ambiguous pair that is almost the same, but we never really define it, so we all use it a little differently...